### ISSUE TITLE
Masked GlobalAveragePooling1D fails when size of step axis is not known statically

### ISSUE THREAD

**Issue by mpdn** [tensorflow/tensorflow#31636]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Mac OS
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 1.14.0, 2.0.0b
- Python version: 3.6.8

**Code to reproduce the issue**

This code:
```
import tensorflow as tf
input = tf.keras.layers.Input([None, 32])
mask = tf.keras.layers.Input([None], dtype=tf.bool)
tf.keras.layers.GlobalAveragePooling1D()(input, mask)
```

Fails with this exception:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/keras/engine/base_layer.py", line 662, in __call__
    outputs = call_fn(inputs, *args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/keras/layers/pooling.py", line 643, in call
    mask = array_ops.reshape(mask, broadcast_shape)
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/ops/gen_array_ops.py", line 7715, in reshape
    "Reshape", tensor=tensor, shape=shape, name=name)
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/framework/op_def_library.py", line 530, in _apply_op_helper
    raise err
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/framework/op_def_library.py", line 527, in _apply_op_helper
    preferred_dtype=default_dtype)
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/framework/ops.py", line 1237, in internal_convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/framework/constant_op.py", line 305, in _constant_tensor_conversion_function
    return constant(v, dtype=dtype, name=name)
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/framework/constant_op.py", line 246, in constant
    allow_broadcast=True)
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/framework/constant_op.py", line 284, in _constant_impl
    allow_broadcast=allow_broadcast))
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/framework/tensor_util.py", line 563, in make_tensor_proto
    "supported type." % (type(values), values))
TypeError: Failed to convert object of type <class 'list'> to Tensor. Contents: [-1, None, 1]. Consider casting elements to a supported type.
```

I would expect masked global average pooling to be no different than regular global average pooling (where unknown step axis is supported) 

I think this is caused by the two lines here: https://github.com/tensorflow/tensorflow/blob/e19c354920c3b246dda6598229210a582caaa1a9/tensorflow/python/keras/layers/pooling.py#L641-L642

Instead of finding the broadcast shape using `as_list` it should use the shape tensor directly and `tf.concat` them together (or whatever is an appropriate equivalent in the keras library).

**Comment by gadagashwini** [comment]
Was able to replicate the issue on Colab with Tensorflow 2.0.0.beta1. Please see the gist [here](https://colab.sandbox.google.com/drive/1AijetZx2ofPuBztdlKovFiC7ZpiT2QFs#scrollTo=l9gPx9dqn4Uu). Thanks!

**Comment by vsimkus** [comment]
Had the same issue when using masking and `GlobalAveragePooling1D` with the Keras `Sequential`. 

For anyone else running into this - it is fixed in `v2.1.0-rc0` and later versions.

https://github.com/tensorflow/tensorflow/blob/9837eceb39171aba9e28dc1f120f53271b6b1ef0/tensorflow/python/keras/layers/pooling.py#L642-L643

**Comment by jvishnuvardhan** [comment]
@mpdn As mentioned by @vsimkus, this issue was resolved in latest version. Please take a look at the [gist](https://colab.sandbox.google.com/gist/jvishnuvardhan/42a3edc1d4fc4e771b0417e5f2b2675c/tf31636.ipynb) for your reference. Thanks!

I am closing this issue as it was resolved. Please feel free to open if the issue persists again. thanks!

**Comment by lzmax888** [comment]
Yes, it is fixed in 2.1, while the issue still persists in 2.0.1.

And if I upgrade to 2.1, the model cannot predict since 'Error when checking model input'.

The issue can be reproduced by 
https://colab.research.google.com/drive/1hMLd5-r82FrnFnBub-B-fVW78Px4KPX1

Anyone can fix it?

Thanks
